File: 42. Trapping Rain Water/brute_force.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        max_left = []
        l = height[0]
        max_left.append(l)
        for i in range(1,n):
            l = max(l,height[i])
            max_left.append(l)
        r = height[n-1]
        max_right = [0]*n
        max_right[n-1] = r
        for i in range(n-2,-1,-1):
            r = max(r,height[i])
            max_right[i] = r
        d = 0
        for i in range(n):
            f = min(max_left[i],max_right[i]) - height[i]
            if f>0:
                d += f
        return d

File: 42. Trapping Rain Water/brute_force.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> max_left,max_right(n);
        int l =height[0];
        max_left.push_back(l);
        
        for(int i=1; i<height.size(); i++){
            l = max(l,height[i]);
            max_left.push_back(l);
        }
        int r = height[height.size()-1];
        max_right[n-1] = r;
        for(int i= height.size()-2; i>=0; i--){
            r = max(r,height[i]);
            max_right[i] = r;
        }
        int d=0;
        for(int i=0; i<height.size(); i++){
            int f = min(max_left[i],max_right[i]) - height[i];
            if(f>0){
                d += f;
            }
        }
        return d;
    }
};
